### PR TITLE
Add trajectory interpolation tests

### PR DESCRIPTION
**Pull request by stefanscherzinger**
Add unit tests for the cartesian_trajectory_interpolation.
This also fixes some possible pitfalls when using this library and fixes a bug in the conversion between spline and Cartesian representation.

### CODE REVIEW

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    EXPECT_EQ(state.w[i], 0);
+    EXPECT_EQ(state.w_dot[i], 0);
+  }
+  EXPECT_EQ(state.q.x(), 0);
+  EXPECT_EQ(state.q.y(), 0);
+  EXPECT_EQ(state.q.z(), 0);
+  EXPECT_EQ(state.q.w(), 1);
+}
+
+TEST(TestCartesianState, EmptyStateHasNormalizedQuaternion)
```
Comment (fmauch) [inline cartesian_trajectory_interpolation/test/cartesian_state_test.cpp:63]: Isn't that test redundant with the one before?

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+TEST(TestCartesianState, RosMessageInitializationYieldsNormalizedQuaternions)
+{
+  cartesian_control_msgs::CartesianTrajectoryPoint init;
+  auto c = CartesianState(init);
+  EXPECT_DOUBLE_EQ(1.0, c.q.norm());
+}
+
+TEST(TestCartesianState, ConversionReturnsInitializingArgument)
+{
+  // Fill some fields with random values.
```
Comment (fmauch) [inline cartesian_trajectory_interpolation/test/cartesian_state_test.cpp]: nitpicking: Those values are arbitrary, but not random ;-)

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  // Note that jerk, posture and time_from_start do not have a representation
+  // in CartesianState.
```
Comment (fmauch) [inline cartesian_trajectory_interpolation/test/cartesian_state_test.cpp]:
```suggestion
  // in CartesianState and are therefore initialized to 0 (by default) in order to make their string representations to be expected the same..
```

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  std::stringstream state_str;
+  init_str << init;
+  state_str << state.toMsg();
+
+  EXPECT_EQ(init_str.str(), state_str.str());
+}
+
+TEST(CartesianState, RotationDifferenceIsPlausible)
+{
+  // Subtraction from zero yields the negative argument
```
Comment (fmauch) [inline cartesian_trajectory_interpolation/test/cartesian_state_test.cpp]: That comment should probably go down to the `EXPECT_...` block. Or maybe even better subtract a from b in order to get positive values?

**Review by fmauch — COMMENTED** [review]
Looks good to me.

However, actual interpolation isn't really tested. Maybe adding a simple trajectory resulting in a straight line between two points might be a good option. Would you think, this would have any benefit @stefanscherzinger ?

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp (hunk 2 above)
Comment (stefanscherzinger) [reply cartesian_trajectory_interpolation/test/cartesian_state_test.cpp]: True, changed.

Comment (stefanscherzinger) [inline cartesian_trajectory_interpolation/test/cartesian_state_test.cpp]:
```suggestion
  // Fill some fields with arbitrary values.
```

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp (hunk 1 above)
Comment (stefanscherzinger) [reply cartesian_trajectory_interpolation/test/cartesian_state_test.cpp:63]: Yes, looks like it. I'll change that.

**Comment by stefanscherzinger** [comment]
@fmauch Concerning the interpolation test: I'm somehow undecided. The core functionality should be tested in the (quintic) spline interpolation we use as library. But you are right, the semantic test for a Cartesian application could be beneficial. I'll add some test later.

**Comment by fmauch** [comment]
> @fmauch Concerning the interpolation test: I'm somehow undecided. The core functionality should be tested in the (quintic) spline interpolation we use as library. But you are right, the semantic test for a Cartesian application could be beneficial. I'll add some test later.

Yes, I was thinking the same, which is why I asked. If its not too much effort, I think it would be nice to have, but definitely not a showstopper.

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp (hunk 4 above)
Comment (stefanscherzinger) [reply cartesian_trajectory_interpolation/test/cartesian_state_test.cpp]: Moved the comment but I would like to keep the subtraction with zero.

File: cartesian_trajectory_interpolation/test/cartesian_state_test.cpp (hunk 1 above)
Comment (stefanscherzinger) [reply cartesian_trajectory_interpolation/test/cartesian_state_test.cpp:63]: Fixed in 41425da5578e4

**Comment by stefanscherzinger** [comment]
Added a test for linear interpolation in afd4604ac6bc11605f3c11a9a0f677bad612a401

**Review by fmauch — APPROVED** [review]
LGTM

**Pull request merged**